File: src/main/java/ltdjms/discord/shop/domain/FiatOrder.java
package ltdjms.discord.shop.domain;

import java.time.Instant;
import java.util.Objects;

import ltdjms.discord.product.domain.Product;

/** Fiat order tracked until payment callback marks it as paid or expired. */
public record FiatOrder(
    Long id,
    long guildId,
    long buyerUserId,
    long productId,
    String productName,
    Product.RewardType fulfillmentRewardType,
    Long fulfillmentRewardAmount,
    boolean fulfillmentAutoCreateEscortOrder,
    String fulfillmentEscortOptionCode,
    String orderNumber,
    String paymentNo,
    long amountTwd,
    Long listPriceTwd,
    Long chargedAmountTwd,
    Status status,
    String tradeStatus,
    String paymentMessage,
    Instant paidAt,
    Instant expireAt,
    Instant expiredAt,
    String terminalReason,
    Instant buyerNotifiedAt,
    Instant rewardGrantedAt,
    Instant fulfilledAt,
    Instant adminNotifiedAt,
    String lastCallbackPayload,
    int reconciliationAttemptCount,
    Instant reconciliationNextAttemptAt,
    Instant createdAt,
    Instant updatedAt) {

  public enum Status {
    PENDING_PAYMENT,
    PAID,
    EXPIRED
  }

  public FiatOrder {
    Objects.requireNonNull(productName, "productName must not be null");
    Objects.requireNonNull(orderNumber, "orderNumber must not be null");
    Objects.requireNonNull(paymentNo, "paymentNo must not be null");
    Objects.requireNonNull(status, "status must not be null");
    Objects.requireNonNull(expireAt, "expireAt must not be null");
    Objects.requireNonNull(createdAt, "createdAt must not be null");
    Objects.requireNonNull(updatedAt, "updatedAt must not be null");

    if (productName.isBlank()) {
      throw new IllegalArgumentException("productName must not be blank");
    }
    if (orderNumber.isBlank()) {
      throw new IllegalArgumentException("orderNumber must not be blank");
    }
    if (paymentNo.isBlank()) {
      throw new IllegalArgumentException("paymentNo must not be blank");
    }
    if (productName.length() > 100) {
      throw new IllegalArgumentException("productName must not exceed 100 characters");
    }
    if ((fulfillmentRewardType == null) != (fulfillmentRewardAmount == null)) {
      throw new IllegalArgumentException(
          "fulfillmentRewardType and fulfillmentRewardAmount must both be specified or both be"
              + " null");
    }
    if (fulfillmentRewardAmount != null && fulfillmentRewardAmount <= 0) {
      throw new IllegalArgumentException("fulfillmentRewardAmount must be positive");
    }
    if (fulfillmentEscortOptionCode != null && fulfillmentEscortOptionCode.length() > 120) {
      throw new IllegalArgumentException(
          "fulfillmentEscortOptionCode must not exceed 120 characters");
    }
    if (fulfillmentAutoCreateEscortOrder) {
      if (fulfillmentEscortOptionCode == null || fulfillmentEscortOptionCode.isBlank()) {
        throw new IllegalArgumentException(
            "fulfillmentEscortOptionCode is required when fulfillmentAutoCreateEscortOrder is"
                + " enabled");
      }
    } else if (fulfillmentEscortOptionCode != null && !fulfillmentEscortOptionCode.isBlank()) {
      throw new IllegalArgumentException(
          "fulfillmentEscortOptionCode requires fulfillmentAutoCreateEscortOrder to be enabled");
    }
    if (orderNumber.length() > 32) {
      throw new IllegalArgumentException("orderNumber must not exceed 32 characters");
    }
    if (paymentNo.length() > 32) {
      throw new IllegalArgumentException("paymentNo must not exceed 32 characters");
    }
    if (amountTwd <= 0) {
      throw new IllegalArgumentException("amountTwd must be positive");
    }
    if (listPriceTwd != null && listPriceTwd <= 0) {
      throw new IllegalArgumentException("listPriceTwd must be positive when provided");
    }
    if (chargedAmountTwd != null && chargedAmountTwd <= 0) {
      throw new IllegalArgumentException("chargedAmountTwd must be positive when provided");
    }
    if (listPriceTwd != null
        && chargedAmountTwd != null
        && chargedAmountTwd > listPriceTwd) {
      throw new IllegalArgumentException("chargedAmountTwd must not exceed listPriceTwd");
    }
    if (status == Status.PENDING_PAYMENT && paidAt != null) {
      throw new IllegalArgumentException("paidAt must be null when status is PENDING_PAYMENT");
    }
    if (status == Status.PAID && paidAt == null) {
      throw new IllegalArgumentException("paidAt is required when status is PAID");
    }
    if (status == Status.EXPIRED) {
      if (paidAt != null) {
        throw new IllegalArgumentException("paidAt must be null when status is EXPIRED");
      }
      if (expiredAt == null) {
        throw new IllegalArgumentException("expiredAt is required when status is EXPIRED");
      }
      if (terminalReason == null || terminalReason.isBlank()) {
        throw new IllegalArgumentException("terminalReason is required when status is EXPIRED");
      }
    } else if (expiredAt != null) {
      throw new IllegalArgumentException("expiredAt must be null unless status is EXPIRED");
    }
    if (terminalReason != null && terminalReason.isBlank()) {
      throw new IllegalArgumentException("terminalReason must not be blank when provided");
    }
  }

  public static FiatOrder createPending(
      long guildId,
      long buyerUserId,
      long productId,
      String productName,
      Product.RewardType fulfillmentRewardType,
      Long fulfillmentRewardAmount,
      boolean fulfillmentAutoCreateEscortOrder,
      String fulfillmentEscortOptionCode,
      String orderNumber,
      String paymentNo,
      long amountTwd,
      Long listPriceTwd,
      Long chargedAmountTwd,
      Instant expireAt) {
    Instant now = Instant.now();
    return new FiatOrder(
        null,
        guildId,
        buyerUserId,
        productId,
        productName,
        fulfillmentRewardType,
        fulfillmentRewardAmount,
        fulfillmentAutoCreateEscortOrder,
        fulfillmentEscortOptionCode,
        orderNumber,
        paymentNo,
        amountTwd,
        listPriceTwd,
        chargedAmountTwd,
        Status.PENDING_PAYMENT,
        null,
        null,
        null,
        expireAt,
        null,
        null,
        null,
        null,
        null,
        null,
        null,
        0,
        null,
        now,
        now);
  }

  public static FiatOrder createPending(
      long guildId,
      long buyerUserId,
      long productId,
      String productName,
      String orderNumber,
      String paymentNo,
      long amountTwd,
      Instant expireAt) {
    return createPending(
        guildId,
        buyerUserId,
        productId,
        productName,
        null,
        null,
        false,
        null,
        orderNumber,
        paymentNo,
        amountTwd,
        amountTwd,
        amountTwd,
        expireAt);
  }

  public boolean hasFulfillmentReward() {
    return fulfillmentRewardType != null && fulfillmentRewardAmount != null;
  }

  public boolean shouldAutoCreateEscortOrder() {
    return fulfillmentAutoCreateEscortOrder
        && fulfillmentEscortOptionCode != null
        && !fulfillmentEscortOptionCode.isBlank();
  }

  public Product toFulfillmentProduct() {
    return new Product(
        productId,
        guildId,
        productName,
        null,
        fulfillmentRewardType,
        fulfillmentRewardAmount,
        null,
        amountTwd,
        fulfillmentAutoCreateEscortOrder,
        fulfillmentEscortOptionCode,
        createdAt,
        updatedAt);
  }

  public boolean isPaid() {
    return status == Status.PAID;
  }

  public boolean isExpired() {
    return status == Status.EXPIRED;
  }

  public boolean isTerminal() {
    return status == Status.PAID || status == Status.EXPIRED;
  }

  public boolean isFulfilled() {
    return fulfilledAt != null;
  }

  public boolean isBuyerNotified() {
    return buyerNotifiedAt != null;
  }

  public boolean isRewardGranted() {
    return rewardGrantedAt != null;
  }

  public boolean isAdminNotified() {
    return adminNotifiedAt != null;
  }
}

File: packages/shop/src/domain/fiat-order.ts
import { z } from 'zod';
import { type Product, type RewardType } from './product-types.js';

// ---- Status enum ----

export enum FiatOrderStatus {
  PENDING_PAYMENT = 'PENDING_PAYMENT',
  PAID = 'PAID',
  EXPIRED = 'EXPIRED',
}

// ---- Zod schema ----

const rewardTypeSchema = z.enum(['CURRENCY', 'TOKEN']).nullable();

export const FiatOrderSchema = z
  .object({
    id: z.number().int().positive().nullable(),
    guildId: z.number(),
    buyerUserId: z.number(),
    productId: z.number(),
    productName: z
      .string()
      .min(1, 'productName must not be blank')
      .max(100, 'productName must not exceed 100 characters'),
    fulfillmentRewardType: rewardTypeSchema,
    fulfillmentRewardAmount: z.number().int().nullable(),
    fulfillmentAutoCreateEscortOrder: z.boolean(),
    fulfillmentEscortOptionCode: z.string().nullable(),
    orderNumber: z
      .string()
      .min(1, 'orderNumber must not be blank')
      .max(32, 'orderNumber must not exceed 32 characters'),
    paymentNo: z
      .string()
      .min(1, 'paymentNo must not be blank')
      .max(32, 'paymentNo must not exceed 32 characters'),
    amountTwd: z.number().int().positive('amountTwd must be positive'),
    status: z.nativeEnum(FiatOrderStatus),
    tradeStatus: z.string().nullable(),
    paymentMessage: z.string().max(512).nullable(),
    paidAt: z.date().nullable(),
    expireAt: z.date(),
    expiredAt: z.date().nullable(),
    terminalReason: z.string().max(128).nullable(),
    buyerNotifiedAt: z.date().nullable(),
    rewardGrantedAt: z.date().nullable(),
    fulfilledAt: z.date().nullable(),
    adminNotifiedAt: z.date().nullable(),
    lastCallbackPayload: z.string().nullable(),
    fulfillmentProcessingAt: z.date().nullable(),
    adminNotificationProcessingAt: z.date().nullable(),
    reconciliationProcessingAt: z.date().nullable(),
    reconciliationAttemptCount: z.number().int().default(0),
    reconciliationNextAttemptAt: z.date().nullable(),
    createdAt: z.date(),
    updatedAt: z.date(),
  })
  .refine(
    (data) => (data.fulfillmentRewardType === null) === (data.fulfillmentRewardAmount === null),
    {
      message:
        'fulfillmentRewardType and fulfillmentRewardAmount must both be specified or both be null',
    },
  )
  .refine(
    (data) => {
      if (data.fulfillmentRewardAmount !== null && data.fulfillmentRewardAmount <= 0) return false;
      return true;
    },
    { message: 'fulfillmentRewardAmount must be positive' },
  )
  .refine(
    (data) => {
      if (data.fulfillmentEscortOptionCode !== null && data.fulfillmentEscortOptionCode.length > 120)
        return false;
      return true;
    },
    { message: 'fulfillmentEscortOptionCode must not exceed 120 characters' },
  )
  .refine(
    (data) => {
      if (data.fulfillmentAutoCreateEscortOrder) {
        return (
          data.fulfillmentEscortOptionCode !== null &&
          data.fulfillmentEscortOptionCode.trim().length > 0
        );
      }
      return true;
    },
    {
      message:
        'fulfillmentEscortOptionCode is required when fulfillmentAutoCreateEscortOrder is enabled',
    },
  )
  .refine(
    (data) => {
      if (
        !data.fulfillmentAutoCreateEscortOrder &&
        data.fulfillmentEscortOptionCode !== null &&
        data.fulfillmentEscortOptionCode.trim().length > 0
      ) {
        return false;
      }
      return true;
    },
    {
      message:
        'fulfillmentEscortOptionCode requires fulfillmentAutoCreateEscortOrder to be enabled',
    },
  )
  .refine(
    (data) => {
      if (data.status === FiatOrderStatus.PENDING_PAYMENT && data.paidAt !== null) return false;
      return true;
    },
    { message: 'paidAt must be null when status is PENDING_PAYMENT' },
  )
  .refine(
    (data) => {
      if (data.status === FiatOrderStatus.PAID && data.paidAt === null) return false;
      return true;
    },
    { message: 'paidAt is required when status is PAID' },
  )
  .refine(
    (data) => {
      if (data.status === FiatOrderStatus.EXPIRED) {
        if (data.paidAt !== null) return false;
        if (data.expiredAt === null) return false;
        if (data.terminalReason === null || data.terminalReason.trim().length === 0) return false;
      }
      return true;
    },
    { message: 'EXPIRED status requires expiredAt, terminalReason, and paidAt must be null' },
  )
  .refine(
    (data) => {
      if (data.status !== FiatOrderStatus.EXPIRED && data.expiredAt !== null) return false;
      return true;
    },
    { message: 'expiredAt must be null unless status is EXPIRED' },
  );

export type FiatOrder = z.infer<typeof FiatOrderSchema>;

// ---- Factory ----

export function createPending(
  guildId: number,
  buyerUserId: number,
  productId: number,
  productName: string,
  fulfillmentRewardType: RewardType | null,
  fulfillmentRewardAmount: number | null,
  fulfillmentAutoCreateEscortOrder: boolean,
  fulfillmentEscortOptionCode: string | null,
  orderNumber: string,
  paymentNo: string,
  amountTwd: number,
  expireAt: Date,
): FiatOrder {
  const now = new Date();
  return FiatOrderSchema.parse({
    id: null,
    guildId,
    buyerUserId,
    productId,
    productName,
    fulfillmentRewardType,
    fulfillmentRewardAmount,
    fulfillmentAutoCreateEscortOrder,
    fulfillmentEscortOptionCode,
    orderNumber,
    paymentNo,
    amountTwd,
    status: FiatOrderStatus.PENDING_PAYMENT,
    tradeStatus: null,
    paymentMessage: null,
    paidAt: null,
    expireAt,
    expiredAt: null,
    terminalReason: null,
    buyerNotifiedAt: null,
    rewardGrantedAt: null,
    fulfilledAt: null,
    adminNotifiedAt: null,
    lastCallbackPayload: null,
    fulfillmentProcessingAt: null,
    adminNotificationProcessingAt: null,
    reconciliationProcessingAt: null,
    reconciliationAttemptCount: 0,
    reconciliationNextAttemptAt: null,
    createdAt: now,
    updatedAt: now,
  });
}

/**
 * Convenience utility that creates a pending fiat order with no reward, no escort.
 * Delegates to createPending with null/default values for reward and escort fields.
 */
export function createPendingSimple(
  guildId: number,
  buyerUserId: number,
  productId: number,
  productName: string,
  orderNumber: string,
  paymentNo: string,
  amountTwd: number,
  expireAt: Date,
): FiatOrder {
  return createPending(
    guildId,
    buyerUserId,
    productId,
    productName,
    null,
    null,
    false,
    null,
    orderNumber,
    paymentNo,
    amountTwd,
    expireAt,
  );
}

// ---- Helper functions ----

export function isPaid(order: FiatOrder): boolean {
  return order.status === FiatOrderStatus.PAID;
}

export function isExpired(order: FiatOrder): boolean {
  return order.status === FiatOrderStatus.EXPIRED;
}

export function isTerminal(order: FiatOrder): boolean {
  return order.status === FiatOrderStatus.PAID || order.status === FiatOrderStatus.EXPIRED;
}

export function isFulfilled(order: FiatOrder): boolean {
  return order.fulfilledAt !== null;
}

export function isBuyerNotified(order: FiatOrder): boolean {
  return order.buyerNotifiedAt !== null;
}

export function isRewardGranted(order: FiatOrder): boolean {
  return order.rewardGrantedAt !== null;
}

export function isAdminNotified(order: FiatOrder): boolean {
  return order.adminNotifiedAt !== null;
}

export function hasFulfillmentReward(order: FiatOrder): boolean {
  return order.fulfillmentRewardType !== null && order.fulfillmentRewardAmount !== null;
}

export function shouldAutoCreateEscortOrder(order: FiatOrder): boolean {
  return (
    order.fulfillmentAutoCreateEscortOrder &&
    order.fulfillmentEscortOptionCode !== null &&
    order.fulfillmentEscortOptionCode.trim().length > 0
  );
}

function isRewardType(value: unknown): value is RewardType {
  return value === 'CURRENCY' || value === 'TOKEN';
}

export function toFulfillmentProduct(order: FiatOrder): Product {
  return {
    id: order.productId,
    guildId: order.guildId,
    name: order.productName,
    description: null,
    rewardType: isRewardType(order.fulfillmentRewardType) ? order.fulfillmentRewardType : null,
    rewardAmount: order.fulfillmentRewardAmount,
    currencyPrice: null,
    fiatPriceTwd: order.amountTwd,
    autoCreateEscortOrder: order.fulfillmentAutoCreateEscortOrder,
    escortOptionCode: order.fulfillmentEscortOptionCode,
    createdAt: order.createdAt,
    updatedAt: order.updatedAt,
  };
}
